### PR TITLE
Reorganize codegen Source Code

### PR DESCRIPTION
**Pull request by apavlo**
Here is a cleaned up version of the codegen reorganization (#690).

1. I added four new directories. Note that I only put `If`, `Loop`, and `VectorizedLoop` under a sub-namespace (i.e., `peloton::codegen::lang`).
   * `expression`
   * `lang`
   * `operator`
   * `proxy`
2. I renamed 'codegen/utils' to 'codegen/util'

https://www.youtube.com/watch?v=fAgevjCJRmI

---------------

### Possible Additions

@pmenon @seojungmin Here are some additional things that we should discuss:

* Do we want to put all of the the sub-classes under their own namespaces too?
* Should we rename the abstract base classes to match the rest of the code?
   * `OperatorTranslator` -> `AbstractOperatorTranslator`
   * `ExpressionTranslator` -> `AbstractExpressionTranslator`
* Why is there both 'codegen/oa_hash_table.cpp' **and** 'codegen/util/oa_hash_table.cpp'?

### CODE REVIEW

File: src/include/codegen/operator/hash_group_by_translator.h [hunk 1]
```diff
@@ -16,7 +16,7 @@
 #include "codegen/compilation_context.h"
 #include "codegen/consumer_context.h"
 #include "codegen/oa_hash_table.h"
-#include "codegen/operator_translator.h"
+#include "operator_translator.h"
```
Comment (pmenon) [inline src/include/codegen/operator/hash_group_by_translator.h]: Change to `#include "codegen/operator/operator_translator.h"`

File: src/include/codegen/operator/hash_join_translator.h [hunk 2]
```diff
@@ -15,7 +15,7 @@
 #include "codegen/compilation_context.h"
 #include "codegen/consumer_context.h"
 #include "codegen/oa_hash_table.h"
-#include "codegen/operator_translator.h"
+#include "operator_translator.h"
```
Comment (pmenon) [inline src/include/codegen/operator/hash_join_translator.h]: Change to `#include "codegen/operator/operator_translator.h"`

File: src/include/codegen/operator/order_by_translator.h [hunk 3]
```diff
@@ -13,7 +13,7 @@
 #pragma once
 
 #include "codegen/compilation_context.h"
-#include "codegen/operator_translator.h"
+#include "operator_translator.h"
```
Comment (pmenon) [inline src/include/codegen/operator/order_by_translator.h]: Change to `#include "codegen/operator/operator_translator.h"`

File: src/include/codegen/operator/projection_translator.h [hunk 4]
```diff
@@ -13,7 +13,7 @@
 #pragma once
 
 #include "codegen/compilation_context.h"
-#include "codegen/operator_translator.h"
+#include "operator_translator.h"
```
Comment (pmenon) [inline src/include/codegen/operator/projection_translator.h]: Change to `#include "codegen/operator/operator_translator.h"`

File: src/include/codegen/operator/table_scan_translator.h [hunk 5]
```diff
@@ -14,7 +14,7 @@
 
 #include "codegen/compilation_context.h"
 #include "codegen/consumer_context.h"
-#include "codegen/operator_translator.h"
+#include "operator_translator.h"
```
Comment (pmenon) [inline src/include/codegen/operator/table_scan_translator.h]: Change to `#include "codegen/operator/operator_translator.h"`

File: src/include/codegen/proxy/deleter_proxy.h [hunk 6]
```diff
@@ -13,8 +13,8 @@
 #pragma once
 
 #include "codegen/codegen.h"
-#include "codegen/data_table_proxy.h"
-#include "codegen/transaction_proxy.h"
+#include "data_table_proxy.h"
+#include "transaction_proxy.h"
```
Comment (pmenon) [inline src/include/codegen/proxy/deleter_proxy.h]: Change to `#include "codegen/proxy/..."`

File: src/include/codegen/operator/global_group_by_translator.h [hunk 7]
```diff
@@ -14,7 +14,7 @@
 
 #include "codegen/aggregation.h"
 #include "codegen/compilation_context.h"
-#include "codegen/operator_translator.h"
+#include "operator_translator.h"
```
Comment (pmenon) [inline src/include/codegen/operator/global_group_by_translator.h]: Change to `#include "codegen/operator/operator_translator.h"`

File: test/codegen/if_test.cpp [hunk 8]
```diff
@@ -13,12 +13,12 @@
 #include "common/harness.h"
 
 #include "codegen/function_builder.h"
-#include "codegen/if.h"
-#include "codegen/loop.h"
-#include "codegen/runtime_functions_proxy.h"
+#include "codegen/lang/if.h"
+#include "codegen/lang/loop.h"
+#include "include/codegen/proxy/runtime_functions_proxy.h"
```
Comment (pmenon) [inline test/codegen/if_test.cpp]: Remove `include` from path.

File: test/codegen/testing_codegen_util.cpp [hunk 9]
```diff
@@ -10,10 +10,10 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include "codegen/codegen_test_util.h"
+#include "codegen/testing_codegen_util.h"
 
-#include "codegen/runtime_functions_proxy.h"
-#include "codegen/values_runtime_proxy.h"
+#include "include/codegen/proxy/runtime_functions_proxy.h"
+#include "include/codegen/proxy/values_runtime_proxy.h"
```
Comment (pmenon) [inline test/codegen/testing_codegen_util.cpp]: Remove `include/` from path.

**Comment by pmenon** [comment]
1. I don't see the benefit of changing names, aside from maybe aligning with the rest of the code. If anything, we should keep `OperatorTransltor` and `ExpressionTranslator` as true interfaces with only pure virtual functions and have abstract subclasses with some helpers.
 
2. Everything under `util` are self-contained data structures (hash tables, sorters etc.). The identically-named classes not under `util` are interfaces that simplify using `util` classes from codegen. They're the bridge that use `proxy` to ease the pain of calling into C++.

File: src/include/codegen/operator/hash_group_by_translator.h (hunk 1 above)
Comment (apavlo) [inline src/include/codegen/operator/hash_group_by_translator.h]: Fixed.

File: src/include/codegen/operator/hash_join_translator.h (hunk 2 above)
Comment (apavlo) [inline src/include/codegen/operator/hash_join_translator.h]: Fixed.

File: src/include/codegen/operator/order_by_translator.h (hunk 3 above)
Comment (apavlo) [inline src/include/codegen/operator/order_by_translator.h]: Fixed.

File: src/include/codegen/operator/projection_translator.h (hunk 4 above)
Comment (apavlo) [inline src/include/codegen/operator/projection_translator.h]: Fixed.

File: src/include/codegen/proxy/deleter_proxy.h (hunk 6 above)
Comment (apavlo) [inline src/include/codegen/proxy/deleter_proxy.h]: Fixed.

File: src/include/codegen/operator/global_group_by_translator.h (hunk 7 above)
Comment (apavlo) [inline src/include/codegen/operator/global_group_by_translator.h]: Fixed.

File: test/codegen/if_test.cpp (hunk 8 above)
Comment (apavlo) [inline test/codegen/if_test.cpp]: Fixed.

File: test/codegen/testing_codegen_util.cpp (hunk 9 above)
Comment (apavlo) [inline test/codegen/testing_codegen_util.cpp]: Fixed.

**Comment by apavlo** [comment]
@pmenon Let me know if you're okay with this and I can merge.

**Comment by pmenon** [comment]
LGTM +1

**Pull request merged**